destination.go

```diff
@@ -133,6 +152,133 @@ func (d *Destination) Write(ctx context.Context, records []sdk.Record) (int, err
 	return len(records), nil
 }
 
+func (d *Destination) sendRecordWithRetries(ctx context.Context, record *opencdcv1.Record) error {
+	d.streamMutex.Lock()
+	if d.stream == nil {
+		// stream would be nil in the case of a failed reconnection
+		err := <-d.errCh
+		return err
+	}
+	err := d.stream.Send(record)
+	if err == io.EOF || status.Code(err) == codes.Unavailable {
+		d.streamMutex.Unlock()
+		err := d.waitForReadyState(ctx, d.conn.GetState())
+		if err != nil {
+			return err
+		}
+		return d.sendRecordWithRetries(ctx, record)
+	}
+	d.streamMutex.Unlock()
+	return err
+}
+
+func (d *Destination) recvAckWithRetries(ctx context.Context) (*pb.Ack, error) {
+	d.streamMutex.Lock()
+	if d.stream == nil {
+		err := <-d.errCh
+		return nil, err
+	}
+	ack, err := d.stream.Recv()
+	if err == io.EOF || status.Code(err) == codes.Unavailable {
+		d.streamMutex.Unlock()
+		err := d.waitForReadyState(ctx, d.conn.GetState())
+		if err != nil {
+			return nil, err
+		}
+		return d.recvAckWithRetries(ctx)
```

Comment:
In case of resending records we will need to be careful about the acks that get sent back, otherwise they could break the connector. So let's say we sent over 5 records, get back 2 acks, the connection drops. When the connection is re-established we send the last 3 records again (we didn't get acks for those yet). Now what about the acks that we expect to get back? We can get back 3, 4, 5 or 6 acks, depending on what happened on the server. Maybe it didn't receive the 3 records in the first try, so we will get back just 3 acks for the re-sent records. Maybe it received only 1 record and then the connection dropped, and it will be able to send back the 1 ack for the old record + the 3 acks for the re-sent records.

The fun begins when the connection drops while re-sending records 🙃 This will be hard to get right and stable, I'm fine with opening an issue and tackling deduplication of acks another time.

Reply:
> So let's say we sent over 5 records, get back 2 acks, then the connection drops. When the connection is re-established we send the last 3 records again

when the connection is re-established, we'd actually continue waiting for acks, we won't send the records again until we get the acks. (because we're not receiving acks in a go routine for now, we send records then wait for acks)
So maybe the whole behavior needs to be changed? I mean, the current behavior does guarantee that this problem won't happen, but, waiting for acks will be the bottleneck until sending new records

Reply:
> when the connection is re-established, we'd actually continue waiting for acks, we won't send the records again until we get the acks.

I was describing the behavior I proposed, where the client would resend records for which it didn't receive acks yet. There's currently no guarantee that we'll receive the acks, as the server doesn't have any retrying logic. And even if it did, it could very well happen that the server pipeline was restarted, so it won't send any acks, it will just wait for new records, while the client will wait for acks.

I'm thinking that we might need a handshake before opening the stream if we want to do retries correctly. Meaning that when the connection is established the server sends back the last position (i.e. the last ack) so the client can figure out where it should begin retrying sending records. There are still edge cases though, but I won't go into detail too much, I think it would be better to tackle this separately and think thoroughly about the interaction between the client and server.

---

TL;DR: currently the code successfully re-establishes the connection and re-opens the stream, but it doesn't _necessarily_ resume processing records. We need to figure that part out, even if it's very simple logic that could deliver some records twice, as long as it's predictable and stable.